File: src/main/java/core/procedures/math/Floor.java
package core.procedures.math;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigRational;
import core.scm.Type;

import java.math.BigDecimal;
import java.math.BigInteger;

public final class Floor extends AFn {

  public Floor() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Type.Real.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "floor";
  }

  @Override
  public Number apply1(Object arg) {
    if (arg instanceof Long || arg instanceof Integer || arg instanceof Short || arg instanceof Byte || arg instanceof BigInteger) {
      return (Number) arg;
    } else if (arg instanceof Double) {
      return Math.floor((Double) arg);
    } else if (arg instanceof Float) {
      return Math.floor((Float) arg);
    } else if (arg instanceof BigDecimal) {
      return ((BigDecimal)arg).setScale(0, BigDecimal.ROUND_DOWN);
    } else if (arg instanceof BigRational) {
      return ((BigRational) arg).floor();
    }
    return Math.floor(((Number)arg).doubleValue());
  }
}

File: src/main/java/core/procedures/math/Floor.kt
package core.procedures.math

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Ratio
import core.scm.Type

import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode
import kotlin.math.floor

class Floor : AFn<Number?, Number>(name = "floor", isPure = true, arity = Exactly(1),
                                   mandatoryArgsTypes = arrayOf(Type.Real::class.java)) {

    override operator fun invoke(arg: Number?) = when (arg) {
        is Long, is Int, is Short, is Byte, is BigInteger -> arg
        is Double     -> floor(arg)
        is Float      -> floor(arg.toDouble())
        is BigDecimal -> {
            val result = arg.setScale(0, RoundingMode.DOWN)
            when {
                arg.scale() > 0 -> result.setScale(1)
                else -> result
            }
        }
        is Ratio   -> arg.floor()
        else          -> floor(arg!!.toDouble())
    }!!
}
